File: ios/Classes/PocketapePlugin.swift
import Flutter
import UIKit

public class PocketapePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    static var shared: PocketapePlugin?
    private var arKitManager: ARKitManager?
    private var flutterEventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = PocketapePlugin()
        shared = instance
        instance.arKitManager = ARKitManager()
        let messenger = registrar.messenger()
        // The _DefaultBinaryMessenger does buffer 1 event, even when no listener is registered.
        // This causes the problem of a race condition when an event was or is added to the buffer while the stream is being canceled:
        // eventBuffer = [EventX];
        // myStream.cancel();
        // myStream.listen(...); -> returns EventX
        //
        // Therefore we disable buffering by setting it to 0
        // https://api.flutter.dev/flutter/dart-ui/ChannelBuffers-class.html
        let bufferChannel = FlutterMethodChannel(name: "ar_events", binaryMessenger: messenger)
        bufferChannel.resizeBuffer(0)

        let eventChannel = FlutterEventChannel(name: "ar_events", binaryMessenger: messenger)
        eventChannel.setStreamHandler(instance)
        instance.arKitManager?.setupSceneView()
    }
  
  func sendEventToFlutter(value: Any) {
    flutterEventSink?(value)
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    flutterEventSink = events
    arKitManager?.startSession()
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    arKitManager?.stopSession()
    flutterEventSink = nil
    return nil
  }
}

File: android/src/main/kotlin/com/openernext/pocketape/PocketapePlugin.kt
package com.openernext.pocketape

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.*

class PocketapePlugin : FlutterPlugin, StreamHandler {

    private lateinit var eventChannel: EventChannel
    private var arCoreManager: ARCoreManager? = null
    private var eventSink: EventSink? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val messenger = flutterPluginBinding.binaryMessenger

        // The _DefaultBinaryMessenger does buffer 1 event, even when no listener is registered.
        // This causes the problem of a race condition when an event was or is added to the buffer while the stream is being canceled:
        // eventBuffer = [EventX];
        // myStream.cancel();
        // myStream.listen(...); -> returns EventX
        //
        // Therefore we disable buffering by setting it to 0
        // https://api.flutter.dev/flutter/dart-ui/ChannelBuffers-class.html
        MethodChannel(messenger, "ar_events").resizeChannelBuffer(0);

        eventChannel = EventChannel(messenger, "ar_events")
        arCoreManager = ARCoreManager(flutterPluginBinding.applicationContext, this)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding)  {
        eventChannel.setStreamHandler(null)
        arCoreManager?.stopSession()
    }

    fun sendEventToFlutter(value: Any) {
        eventSink?.success(value)
    }

    override fun onListen(arguments: Any?, events: EventSink?) {
        eventSink = events
        arCoreManager?.startSession()
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        arCoreManager?.stopSession()
    }
}